ide/coqide/coqOps.ml

```diff
@@ -138,11 +144,27 @@ end = struct
 end
 open SentenceId
 
+(** move iter by the number of chars used to represent the following
+    n_bytes.  Used to adjust locations in case the sentence is moved after it's
+    sent to Coq in an "add" call. *)
+let b2c (buffer : GText.buffer) n_bytes iter0 =
+  let rec cvt iter len =
+    if len <= 0 then
+      iter
+    else
+      cvt iter#forward_char (len - (Ideutils.ulen iter#char))
+  in
+  cvt iter0 n_bytes
+
 (* Given a Coq loc, convert it to a pair of iterators start / end in
    the buffer. *)
-let coq_loc_to_gtk_offset (buffer : GText.buffer) loc =
-  buffer#get_iter_at_byte ~line:loc.Loc.line_nb (loc.bp - loc.bol_pos),
-  buffer#get_iter_at_byte ~line:loc.Loc.line_nb_last (loc.ep - loc.bol_pos_last)
+let coq_loc_to_gtk_offset (buffer : GText.buffer) loc sentence =
+  (* n_bytes is invariant even if the sentence was moved though the
+     buffer offset will differ *)
+  let adjust byte_off =
```

Comment:
The whole approach here seems bugged to me, the GTK editor component already keeps track of relocation right? If so, we should not have to do this process ourselves.

Reply:
As I say in the description, "[This] handles the case where the position of a sentence changes after it's been sent to the server (described [here](https://github.com/coq/coq/issues/17023#issuecomment-1367510209))."

When you send a sentence to the server, the bp/ep values for returned Locs in that sentence are based on (relative to) the original bp that was passed in the `add` call.  If the sentence moves, converting the new GTK position to bp will be different than the bp sent in the `add` command and thus incorrect.  Adding the line number to the protocol won't help--you need to remember the original value.

Reply:
What's the invariant that must hold regarding `bp` and `start` in a sentence?

Reply:
Suppose you send a sentence to the server in an "add" request with sentence.bp = 100 and sentence.ep = 200.  The returned loc (say bp=110, ep=120) for a warning will be a subrange of 100..200.  These offsets won't change even if the sentence moves in the buffer.

If the user later enters or removes characters before the sentence and a similar warning is received for the sentence, the 10 and 20 byte offsets relative to sentence.start are still correct.  All that's needed is to figure out how many GTK chars correspond to the first 10 or 20 bytes offset of the sentence, i.e. by examining the first 10 or 20 bytes of chars after `start`.

Maybe "invariant" is not the ideal word.

I think it's pretty straightforward once you grasp it.  More words may just confuse. Probably better explained with a picture but not so easy to create one.  Hope this helps, let me know.  _Tuum passibus variari potest._